syscalls_windows.go

```diff
@@ -0,0 +1,171 @@
+// +build windows
+
+// To use it with windows, you need a tap driver installed on windows.
+// https://github.com/OpenVPN/tap-windows6
+// or just install OpenVPN
+// https://github.com/OpenVPN/openvpn
+package water
+
+import (
+	"errors"
+	"net"
+	"os"
+	"syscall"
+
+	"golang.org/x/sys/windows/registry"
+)
+
+var (
+	IfceNameNotFound  = errors.New("Failed to find the name of interface.")
+	TapDeviceNotFound = errors.New("Failed to find the tap device in registry.")
+	// Device Control Codes
+	tap_win_ioctl_get_mac             = tap_control_code(1, 0)
+	tap_win_ioctl_get_version         = tap_control_code(2, 0)
+	tap_win_ioctl_get_mtu             = tap_control_code(3, 0)
+	tap_win_ioctl_get_info            = tap_control_code(4, 0)
+	tap_ioctl_config_point_to_point   = tap_control_code(5, 0)
+	tap_ioctl_set_media_status        = tap_control_code(6, 0)
+	tap_win_ioctl_config_dhcp_masq    = tap_control_code(7, 0)
+	tap_win_ioctl_get_log_line        = tap_control_code(8, 0)
+	tap_win_ioctl_config_dhcp_set_opt = tap_control_code(9, 0)
+	tap_ioctl_config_tun              = tap_control_code(10, 0)
+	// w32 api
+	file_device_unknown = uint32(0x00000022)
+)
+
+func ctl_code(device_type, function, method, access uint32) uint32 {
+	return (device_type << 16) | (access << 14) | (function << 2) | method
+}
+
+func tap_control_code(request, method uint32) uint32 {
+	return ctl_code(file_device_unknown, request, method, 0)
+}
+
+// getdeviceid finds out a TAP device from registry, it requires privileged right.
+func getdeviceid() (string, error) {
+	// TAP driver key location
+	regkey := `SYSTEM\CurrentControlSet\Control\Class\{4D36E972-E325-11CE-BFC1-08002BE10318}`
+	k, err := registry.OpenKey(registry.LOCAL_MACHINE, regkey, registry.ALL_ACCESS)
+	if err != nil {
+		return "", err
+	}
+	defer k.Close()
+	// read all subkeys
+	keys, err := k.ReadSubKeyNames(-1)
+	if err != nil {
+		return "", err
+	}
+	// find the one with ComponentId == "tap0901"
+	for _, v := range keys {
+		key, err := registry.OpenKey(registry.LOCAL_MACHINE, regkey+"\\"+v, registry.ALL_ACCESS)
+		if err != nil {
+			continue
+		}
```

Comment:
I think if you do something like this here, you can avoid all the `goto next` in this loop:

```
defer func(k registry.Key) {
    k.Close()
}(key)
```